sweepbatcher/sweep_batch.go

```diff
@@ -0,0 +1,1369 @@
+		return func() {}, ErrBatchShuttingDown
+	}
+
+	return func() {
+		b.callLeave <- struct{}{}
+	}, nil
+}
+
+// NewBatch creates a new batch.
+func NewBatch(cfg batchConfig, bk batchKit) *batch {
+	return &batch{
+		// We set the ID to a negative value to flag that this batch has
+		// never been persisted, so it needs to be assigned a new ID.
+		id:               -1,
+		state:            Open,
+		sweeps:           make(map[lntypes.Hash]sweep),
+		blockEpochChan:   make(chan int32),
+		spendChan:        make(chan *chainntnfs.SpendDetail),
+		confChan:         make(chan *chainntnfs.TxConfirmation, 1),
+		reorgChan:        make(chan struct{}, 1),
+		errChan:          make(chan error, 1),
+		callEnter:        make(chan struct{}),
+		callLeave:        make(chan struct{}),
+		quit:             make(chan struct{}),
+		completeChan:     bk.completeChan,
+		batchTxid:        bk.batchTxid,
+		wallet:           bk.wallet,
+		chainNotifier:    bk.chainNotifier,
+		signerClient:     bk.signerClient,
+		muSig2SignSweep:  bk.musig2SignSweep,
+		verifySchnorrSig: bk.verifySchnorrSig,
+		purger:           bk.purger,
+		store:            bk.store,
+		cfg:              &cfg,
+	}
+}
+
+// NewBatchFromDB creates a new batch that already existed in storage.
+func NewBatchFromDB(cfg batchConfig, bk batchKit) *batch {
+	return &batch{
+		id:               bk.id,
+		state:            bk.state,
+		primarySweepID:   bk.primaryID,
+		sweeps:           bk.sweeps,
+		blockEpochChan:   make(chan int32),
+		spendChan:        make(chan *chainntnfs.SpendDetail),
+		confChan:         make(chan *chainntnfs.TxConfirmation, 1),
+		reorgChan:        make(chan struct{}, 1),
+		errChan:          make(chan error, 1),
+		callEnter:        make(chan struct{}),
+		callLeave:        make(chan struct{}),
+		quit:             make(chan struct{}),
+		completeChan:     bk.completeChan,
+		batchTxid:        bk.batchTxid,
+		batchPkScript:    bk.batchPkScript,
+		rbfCache:         bk.rbfCache,
+		wallet:           bk.wallet,
+		chainNotifier:    bk.chainNotifier,
+		signerClient:     bk.signerClient,
+		muSig2SignSweep:  bk.musig2SignSweep,
+		verifySchnorrSig: bk.verifySchnorrSig,
+		purger:           bk.purger,
+		store:            bk.store,
+		log:              bk.log,
+		cfg:              &cfg,
+	}
+}
+
+// addSweep tries to add a sweep to the batch. If this is the first sweep being
+// added to the batch then it also sets the primary sweep ID.
+func (b *batch) addSweep(ctx context.Context, sweep *sweep) (bool, error) {
+	done, err := b.scheduleNextCall()
+	defer done()
+
+	if err != nil {
+		return false, err
+	}
+
+	// If the provided sweep is nil, we can't proceed with any checks, so
+	// we just return early.
+	if sweep == nil {
+		return false, nil
+	}
+
+	// Before we run through the acceptance checks, let's just see if this
+	// sweep is already in our batch. In that case, just update the sweep.
+	_, ok := b.sweeps[sweep.swapHash]
+	if ok {
+		b.sweeps[sweep.swapHash] = *sweep
```

Comment:
So here we essentially "attach" the sweep coming from a resumed swap if the sweeper resumed it before the swap added it. Meaning we add the sweep notifier. Perhaps worth a mention.

Reply:
Yes since we already have that sweep added, at this point we just attach it again. :+1: